extensions/cl_ext_buffer_device_address.asciidoc

```diff
@@ -0,0 +1,320 @@
+data structures that need pointer references to objects.
+
+Shared Virtual Memory (SVM) introduced in OpenCL 2.0 is the first feature
+that enables raw device side pointers in the OpenCL standard. Its coarse-grain
+variant is relatively simple to implement on various platforms in terms of
+coherency requirements, but it requires mapping the buffer's address range
+to the host virtual address space although it might not be needed by the
+application. This is not an issue in systems which can provide virtual memory
+across the platform, but might provide implementation challenges in cases
+where the device presents a global memory with a disjoint address space
+(that can also be a physical memory address space) or, for example, when
+a barebone embedded system lacks virtual memory support altogether.
+
+Various higher-level APIs present a memory allocation routine which can
+allocate device-only memory and provide raw pointers to it without guarentees
+of system-wide uniqueness: For example, minimal implementations of OpenMP's
+omp_target_alloc() and CUDA/HIP's cudaMalloc()/hipMalloc() do not require a shared
+address space between the host and the device. This extension is meant to
+provide a minimal set of features to implement such APIs using the cl_mem
+buffers without requiring a shared virtual address space between the host and
+the device.
+
+=== New API Function
+
+include::{generated}/api/protos/clSetKernelArgDevicePointerEXT.txt[]
+
+=== New API Enums
+
+Enums for enabling device pointer properties when creating a buffer
+{clCreateBuffer}, see <<clCreateBuffer, the list of supported memory flag values table>>:
+
+[source]
+----
+#define CL_MEM_DEVICE_SHARED_ADDRESS_EXT (1ul << 31)
+#define CL_MEM_DEVICE_PRIVATE_ADDRESS_EXT (1ul << 30)
+----
+
+Enums for querying the device pointer from the cl_mem <<clGetMemObjectInfo, the list of supported param_names table>>:
+
+[source]
+----
+#define CL_MEM_DEVICE_PTR_EXT 0xff01
+----
+
+Enums for setting information of indirect device pointer accesses to kernels <<clSetKernelExecInfo, the list of supported param_names table>>. This is for OpenCL 2.0 and above. When implementing the
+extension on an older OpenCL version, indirect device pointer access is not supported.
+
+[source]
+----
+#define CL_KERNEL_EXEC_INFO_DEVICE_PTRS_EXT 0x11B8
+----
+
+== New API Types
+
+Returned as the query result value *clGetMemObjectInfo* with `CL_DEVICE_PTR_EXT`.
+
+[source]
+----
+typedef cl_ulong cl_mem_device_address_EXT;
+----
+
+Returned as the query result value *clGetMemObjectInfo* with `CL_DEVICE_PTRS_EXT`.
+
+[source]
+----
+typedef struct _cl_mem_device_address_pair_EXT
+{
+  cl_device_id device;
+  cl_mem_device_address_EXT address;
+} cl_mem_device_address_pair_EXT;
+----
+
+== Modifications to the OpenCL API Specification
+
+=== Section 5.2.1 - Creating Buffer Objects:
+
+Add new allocation flags <<clCreateBuffer, List of supported memory flag values table>>:
```

Comment:
flags are in relatively short supply and it seems this feature would only ever apply to buffers and likely be superseded by the upcoming USM extension. Have you considered using buffer properties instead (this approach would have the drawback of requiring OpenCL 3.0)?

Reply:
I wonder if USM can be implemented on top of Vulkan effectively. I haven't really checked in greater detail how much of the SVM semantics are required by USM. But the main issue is that the vulkan driver can allocate GPU vm addresses as they want, which makes it hard to guarantee that allocations can be accessible on the host with the same address.

Reply:
It's not possible to guarantee support on Vulkan currently. I proposed https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/6024 as a cheap of supporting coarse-grain semantics on Vulkan but have not had time to chase this recently.

Reply:
mhhh, but that won't work with multiple devices.

Maybe I should chime in and share what kind of API I settled with within mesa to implement SVM semantics through a a more generic API, not relying on drivers doing much.

The tldr: the consumer can reserve VM ranges and assign addresses to memory allocations directly instead of the driver doing that itself. Most Vulkan drivers do VMA assignment in userspace anyway, so it's not too hard to add support for that.

And drivers not doing Usermode vm management should probably move towards that regardless.

That also allows to e.g. reserve the same range across multiple devices and allow implementing SVM supporting multiple devices across vendors.

Reply:
Yes, that's a limitation of my proposal, single device only. TBH, I see this proposal more as a stop-gap, I suggested it because it's very cheap for Arm (and likely others) to support and provides a strong guarantees in some useful scenarios.

What's hard with the approach you followed is to guarantee that you can get matching ranges on all devices and the host. It might work well in practice though. I was surprised to see how far I got when I prototyped SVM in clvk just using a shadow host allocation created with `mmap` at the same address as the device allocation. Horrible by design but it works on several devices (tm).

Reply:
yeah.. I Intel's SVM implementation just hopes for the best and doesn't even reserve any memory on the host.

Large part of the host VM ranges never gets used, unless you use something like libasan. But the only reason I made drivers advertise those ranges is simply because drivers in mesa use certain ranges for private allocations. `iris` (Intel) generally uses the first 2 or 3 GiB, so there is plenty of space. And I also exclude the ranges where address canonicalization might be needed (high bits set). So that still gives a huge amount of space across most drivers.

Reply:
Changed to a clCreateBufferWithProperties() property.